arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt

```diff
@@ -246,55 +234,72 @@ public fun <K, A, B> Map<K, A>.flatMap(f: (Map.Entry<K, A>) -> Map<K, B>): Map<K
     }
   }
 
+@Deprecated(
+  "Traverse for Either is being deprecated in favor of Either DSL + Map.mapValues.\n$NicheAPI",
+  ReplaceWith(
+    "let<Map<K, A>, Either<E, Map<K, B>>> { m -> either<E, Map<K, B>> { m.mapValues<K, A, B> { (_, a) -> f(a).bind<B>() } } }",
+    "arrow.core.raise.either"
+  )
+)
 @OptIn(ExperimentalTypeInference::class)
 @OverloadResolutionByLambdaReturnType
-public inline fun <K, E, A, B> Map<K, A>.traverse(f: (A) -> Either<E, B>): Either<E, Map<K, B>> {
-  val acc = mutableMapOf<K, B>()
-  forEach { (k, v) ->
-    when (val res = f(v)) {
-      is Right -> acc[k] = res.value
-      is Left -> return@traverse res
-    }
-  }
-  return acc.right()
-}
+public inline fun <K, E, A, B> Map<K, A>.traverse(f: (A) -> Either<E, B>): Either<E, Map<K, B>> =
```

Comment:
Yes, I've seen the issue with `Either.Right<Any>` in multiple places (in our tests)😞 I think it's due to _covariance_ and `Nothing` in the data class constructor. I think this is why Kotlin Std **always** requires specifying the arguments, and `Nothing` is never used anywhere.

If this constructor was `operator fun fun <A, B> Either.Right.Companion.invoke(b: B): Either<A, B> = ...` I think it would not have been a problem, but it comes with some other trade-offs like not being able to call `::Right`.

That being said, I don't think a lot of people are using this pattern in real code, since `traverse` with only `Right` is `map` without `Either`.. In case that the compiler can know `E` and `A` I don't think this `ReplaceWith` results in any problems. So maybe we can leave this as _best effort_?

Reply:
Sure thing. I think we can leave as it is and be *reactive* in the case some raises an issue.